crates/uv-auth/src/keyring.rs

```diff
@@ -17,20 +27,84 @@ pub struct KeyringProvider {
 
 #[derive(Debug)]
 pub(crate) enum KeyringProviderBackend {
-    /// Use the `keyring` command to fetch credentials.
+    /// Use system keyring integration to fetch credentials.
+    Native,
+    /// Use the external `keyring` command to fetch credentials.
     Subprocess,
     #[cfg(test)]
     Dummy(Vec<(String, &'static str, &'static str)>),
 }
 
 impl KeyringProvider {
+    /// Create a new [`KeyringProvider::Native`].
+    pub fn native() -> Self {
+        Self {
+            backend: KeyringProviderBackend::Native,
+        }
+    }
+
     /// Create a new [`KeyringProvider::Subprocess`].
     pub fn subprocess() -> Self {
         Self {
             backend: KeyringProviderBackend::Subprocess,
         }
     }
 
+    /// Store credentials for the given [`Url`] to the keyring if the
+    /// keyring provider backend is `Native`.
+    #[instrument(skip_all, fields(url = % url.to_string(), username))]
+    pub fn store_if_native(&self, url: &DisplaySafeUrl, credentials: &Credentials) {
```

Comment:
We could extend this to the Python `keyring` as well, but for this PR I've limited the scope to the new `native` backend.